### ISSUE TITLE
[BE] Avoid .data usage in FSDP buffer casting

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#96372]
### 🐛 Describe the bug

FSDP casts buffers for mixed precision, but uses `buf.data` to assign to, we should avoid the .data usage by de-registering the old buffer and re-registering the low precision buffer.

### Versions

main

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by awgu** [comment]
Is he upside just to get rid of one unnecessary `.data` usage, or is there a reason we should modify the variable in-place?

**Comment by zhaojuanmao** [comment]
yeah, did we already see some issues using .data for buffers for now?
